### DISCUSSION TITLE
OIDC-based package signing

### DISCUSSION THREAD

**Discussion by lann** [bytecodealliance/SIG-Registries#13] [General]
In the [What is a signature?](https://github.com/bytecodealliance/SIG-Registries/discussions/7) discussion thread, [@carlsverre suggests](https://github.com/bytecodealliance/SIG-Registries/discussions/7#discussioncomment-2703351) investigating sigstore's services, including [Fulcio](https://github.com/sigstore/fulcio), which is a semi-centralized system that authenticates users via OpenID Connect and issues short-lived (e.g. 20 minute) certificates for the purpose of code signing. @squillace [expresses concern](https://github.com/bytecodealliance/SIG-Registries/discussions/7#discussioncomment-2741033) about both the centralization and security implications of using this type of system.

I'd like to discuss whether we would want to support an OIDC PKI system like Fulcio. I see two separate questions:
1. Should we add a user story to support this type of system? From a technical perspective I think this boils down to supporting Web PKI for signing.
2. Should we attempt to use Fulcio (or some similar system, though I'm not aware of any other appealing options) in the code this group produces?

**Comment by squillace** [comment]
Today sigstore argued that they should be "the" public service for the US government. https://blog.chainguard.dev/sigstore-statement-for-openssf-call/. Like I said, I'm quite sure the project or parts of it _will be great in this role_. The idea that any community should predeclare on a single megaproject is for me very, very strange. I would also note that Kubernetes, which was the same initially, succeeded not because it was great but because a strong, diverse community AND other vendors rallied around it. It was, by design, extremely inclusive.

And I still love fulcio and think it and cosign should be a couple of the tools that ARE used here. :-)

**Comment by squillace** [comment]
Oh, per @lann's OP here, I _support_ using fulcio to test things and as a first implementation of signing if we want. I still think that solution is a great one for about 80% of the contributor base. I just think it can't be the only thing people can use.

**Reply by lann** [reply]
Fulcio uses X.509/PKIX, so from that perspective there would be flexibility in supporting other kinds of CAs. One possible design direction would be to allow [namespaces](https://github.com/bytecodealliance/SIG-Registries/discussions/8) or even individual packages to be configured by maintainers with different (self-signed) CAs, with a public Fulcio CA as just one option. The biggest drawback here would be the inherent complexity of PKIX, which shouldn't be ignored.

**Comment by squillace** [comment]
Yeah, we should standardize on the format(s) and just dive into that mess head on. I would expect fulcio to be one of the neato uxes there, no question in my mind.
